// typescript-to-lua/dist/transformation/visitors/binary-expression/bit.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBitOperator = void 0;
exports.transformBinaryBitOperation = transformBinaryBitOperation;
exports.transformUnaryBitOperation = transformUnaryBitOperation;
const ts = __importStar(require("typescript"));
const CompilerOptions_1 = require("../../../CompilerOptions");
const lua = __importStar(require("../../../LuaAST"));
const utils_1 = require("../../../utils");
const diagnostics_1 = require("../../utils/diagnostics");
const isBitOperator = (operator) => operator in bitOperatorToLibOperation;
exports.isBitOperator = isBitOperator;
const bitOperatorToLibOperation = {
    [ts.SyntaxKind.AmpersandToken]: "band",
    [ts.SyntaxKind.BarToken]: "bor",
    [ts.SyntaxKind.CaretToken]: "bxor",
    [ts.SyntaxKind.LessThanLessThanToken]: "lshift",
    [ts.SyntaxKind.GreaterThanGreaterThanToken]: "arshift",
    [ts.SyntaxKind.GreaterThanGreaterThanGreaterThanToken]: "rshift",
};
function transformBinaryBitLibOperation(node, left, right, operator, lib) {
    const functionName = bitOperatorToLibOperation[operator];
    return lua.createCallExpression(lua.createTableIndexExpression(lua.createIdentifier(lib), lua.createStringLiteral(functionName)), [left, right], node);
}
function transformBitOperatorToLuaOperator(context, node, operator) {
    switch (operator) {
        case ts.SyntaxKind.BarToken:
            return lua.SyntaxKind.BitwiseOrOperator;
        case ts.SyntaxKind.CaretToken:
            return lua.SyntaxKind.BitwiseExclusiveOrOperator;
        case ts.SyntaxKind.AmpersandToken:
            return lua.SyntaxKind.BitwiseAndOperator;
        case ts.SyntaxKind.LessThanLessThanToken:
            return lua.SyntaxKind.BitwiseLeftShiftOperator;
        case ts.SyntaxKind.GreaterThanGreaterThanToken:
            context.diagnostics.push((0, diagnostics_1.unsupportedRightShiftOperator)(node));
            return lua.SyntaxKind.BitwiseRightShiftOperator;
        case ts.SyntaxKind.GreaterThanGreaterThanGreaterThanToken:
            return lua.SyntaxKind.BitwiseRightShiftOperator;
    }
}
function transformBinaryBitOperation(context, node, left, right, operator) {
    switch (context.luaTarget) {
        case CompilerOptions_1.LuaTarget.Universal:
        case CompilerOptions_1.LuaTarget.Lua50:
        case CompilerOptions_1.LuaTarget.Lua51:
            context.diagnostics.push((0, diagnostics_1.unsupportedForTarget)(node, "Bitwise operations", context.luaTarget));
            return transformBinaryBitLibOperation(node, left, right, operator, "bit");
        case CompilerOptions_1.LuaTarget.LuaJIT:
            return transformBinaryBitLibOperation(node, left, right, operator, "bit");
        case CompilerOptions_1.LuaTarget.Lua52:
            return transformBinaryBitLibOperation(node, left, right, operator, "bit32");
        default:
            // Lua 5.3+ `>>` is arithmetic (sign-extending), but TS `>>>` is logical (zero-fill).
            // Emit `(left & 0xFFFFFFFF) >> right` to convert to unsigned 32-bit first.
            if (operator === ts.SyntaxKind.GreaterThanGreaterThanGreaterThanToken) {
                const mask = lua.createBinaryExpression(left, lua.createNumericLiteral(0xffffffff, node), lua.SyntaxKind.BitwiseAndOperator, node);
                return lua.createBinaryExpression(lua.createParenthesizedExpression(mask, node), right, lua.SyntaxKind.BitwiseRightShiftOperator, node);
            }
            const luaOperator = transformBitOperatorToLuaOperator(context, node, operator);
            return lua.createBinaryExpression(left, right, luaOperator, node);
    }
}
function transformUnaryBitLibOperation(node, expression, operator, lib) {
    let bitFunction;
    switch (operator) {
        case lua.SyntaxKind.BitwiseNotOperator:
            bitFunction = "bnot";
            break;
        default:
            (0, utils_1.assertNever)(operator);
    }
    return lua.createCallExpression(lua.createTableIndexExpression(lua.createIdentifier(lib), lua.createStringLiteral(bitFunction)), [expression], node);
}
function transformUnaryBitOperation(context, node, expression, operator) {
    switch (context.luaTarget) {
        case CompilerOptions_1.LuaTarget.Universal:
        case CompilerOptions_1.LuaTarget.Lua50:
        case CompilerOptions_1.LuaTarget.Lua51:
            context.diagnostics.push((0, diagnostics_1.unsupportedForTarget)(node, "Bitwise operations", context.luaTarget));
            return transformUnaryBitLibOperation(node, expression, operator, "bit");
        case CompilerOptions_1.LuaTarget.LuaJIT:
            return transformUnaryBitLibOperation(node, expression, operator, "bit");
        case CompilerOptions_1.LuaTarget.Lua52:
            return transformUnaryBitLibOperation(node, expression, operator, "bit32");
        default:
            return lua.createUnaryExpression(expression, operator, node);
    }
}
